ckanext/dgu/bin/set_missing_resource_formats.py

```diff
@@ -0,0 +1,113 @@
+"""
+For those resources that do not have valid formats (i.e. empty, or just whitespace), this
+script will set the value from the QA object that was previously created.
+
+This will trigger an archive/packagezip/qa cycle, so should not be used too frequently.
+"""
+from sqlalchemy.exc import IntegrityError
+from pprint import pprint
+import ckan.plugins.toolkit as t
+from optparse import OptionParser
+
+from ckan.logic import ValidationError
+from ckanext.dgu.bin.common import get_ckanapi
+from running_stats import Stats
+
+stats = Stats()
+ds_stats = Stats()
+
+class SetResourceFormatsCommand(object):
+
+    def __init__(self, config_or_url):
+        self.ckan = get_ckanapi(config_or_url)
+
+    def update_resource_dict(self, resource):
+        """ Set the format on the resource to the format determined by QA """
+        try:
+            qa_info = self.ckan.action.qa_resource_show(id=resource['id'])
+            if qa_info.get('format', ''):
+                resource['format'] = qa_info['format']
+                stats.add("Updating format to %s" % qa_info['format'], resource['id'])
+        except t.ObjectNotFound:
+            # No QA yet
+            return False
+
+        # Resource was changed
+        return True
+
+    def run(self, options):
+        """ Iterate over datasets and process the resources """
+
+        # Create a function to get a list of dataset names to work with
+        get_datasets_fn = self.ckan.action.package_list
+        if options.dataset:
+            get_datasets_fn = lambda: [options.dataset]
+
+        if not options.write:
+            print "NOT writing package as -w was not specified"
+
+
+        # For each dataset, get resources and process those with no format.
+        for pkg_name in get_datasets_fn():
+            updated = False
+
+            pkg = self.ckan.action.package_show(id=pkg_name)
+            resources = pkg['resources']
+            if 'individual_resources':
+                resources = pkg.get('individual_resources', []) + \
+                    pkg.get('timeseries_resources', []) + \
+                    pkg.get('additional_resources', [])
+                del pkg['resources']
+
+            for resource in resources:
+                del resource['revision_id']
+                if resource['format'].strip() == '':
+                    if self.update_resource_dict(resource):
+                        updated = True
+                        resource['format']
+
+
+            # Removing the codelist or schema here does remove it during package_update
+            # Removing last_major_modification is calculated on write.
+            for k in ['last_major_modification', 'schema', 'codelist']:
+                if k in pkg:
+                    del pkg[k]
```

Comment:
There's a missing 'not' in the comment.  Not having them in the dict does not remove them with package_update.